examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+        SolverCG<>    cg (solver_control);
+
+        dst.block(0) = 0.0;
+        cg.solve(system_matrix.block(0,0),
+                 dst.block(0),
+                 utmp,
+                 preconditioner_A);
+
+        n_iterations_A += solver_control.last_step();
+      }
+    else
+      {
+        preconditioner_A.vmult (dst.block(0), utmp);
+        n_iterations_A += 1;
+      }
+  }
+
+  // @sect3{The StokesProblem class}
+  //
+  // This is the main class of the problem.
+  template <int dim>
+  class StokesProblem
+  {
+  public:
+    StokesProblem (const unsigned int degree, SolverType::type solver_type);
+    void run ();
+
+  private:
+    void setup_dofs ();
+    void assemble_system ();
+    void assemble_multigrid ();
+    void solve ();
+    void compute_errors ();
+    void output_results (const unsigned int refinement_cycle) const;
+
+    const unsigned int   degree;
+    SolverType::type     solver_type;
+
+    Triangulation<dim>   triangulation;
+    FESystem<dim>        fe;
+    FESystem<dim>        velocity_fe;
+    DoFHandler<dim>      dof_handler;
+    DoFHandler<dim>      velocity_dof_handler;
+
+    ConstraintMatrix     constraints;
+
+    BlockSparsityPattern      sparsity_pattern;
+    BlockSparseMatrix<double> system_matrix;
+    SparseMatrix<double> pressure_mass_matrix;
+
+    BlockVector<double> solution;
+    BlockVector<double> system_rhs;
+
+    MGLevelObject<SparsityPattern>        mg_sparsity_patterns;
+    MGLevelObject<SparseMatrix<double> >  mg_matrices;
+    MGLevelObject<SparseMatrix<double> >  mg_interface_matrices;
+    MGConstrainedDoFs                     mg_constrained_dofs;
+
+    TimerOutput computing_timer;
+  };
+
+
+
+  template <int dim>
+  StokesProblem<dim>::StokesProblem (const unsigned int degree, SolverType::type solver_type)
+    :
+    degree (degree),
+    solver_type (solver_type),
+    triangulation (Triangulation<dim>::maximum_smoothing),
+    fe (FE_Q<dim>(degree+1), dim, // Finite element for whole system
+        FE_Q<dim>(degree), 1),
+    velocity_fe (FE_Q<dim>(degree+1), dim), // Finite element for velocity-only
+    dof_handler (triangulation),
+    velocity_dof_handler (triangulation),
+    computing_timer (std::cout, TimerOutput::summary,
+                     TimerOutput::wall_times)
+  {}
+
+
+
+// @sect4{StokesProblem::setup_dofs}
+
+// This function sets up things differently based on if you want to use ILU or GMG as a preconditioner.
+  template <int dim>
+  void StokesProblem<dim>::setup_dofs ()
+  {
+    TimerOutput::Scope scope(computing_timer, "Setup");
+
+    system_matrix.clear ();
+    pressure_mass_matrix.clear ();
+
+    // We don't need the multigrid dofs for whole problem finite element
+    dof_handler.distribute_dofs(fe);
+
+    // This first creates and array (0,0,1) which means that it first does everything with index 0 and then 1
+    std::vector<unsigned int> block_component (dim+1,0);
+    block_component[dim] = 1;
```

Comment:
"This" what? Maybe "In the following code block, we first create an array of length dim+1 that is initialized to all zeros; we then set the pressure vector component to 1."

But it's not clear from the comment what the *purpose* of this variable is going to be, and that would be useful to know.